### ISSUE TITLE
Unable to use gnu mpz types for NumberIntegerType

### ISSUE THREAD

**Issue by eric-wieser** [nlohmann/json#3578]
### Description

I was hoping it would be possible to set `NumberIntegerType` and `NumberUnsignedType` (putting aside #2573 for now) to an arbitrary precision integer type ("big integers"), such as on wrapper types around the gnu [multiple precision arithmetic library](https://gmplib.org/).

Unfortunately, this is not permitted as `json_value` is a C-style tagged union that doesn't permit any of its types to have non-trivial destructors/move constructors etc; and `mpz` is stored on the heap. 

I understand this is quite a big ask, but I thought it was worth having an issue explicitly mentioning big integers since I wasn't able to find an existing one via search.

### Reproduction steps

Attempt to use some numeric heap type for the numeric type parameters. I'm afraid I don't have a trivial example to hand without external dependencies.

### Expected vs. actual results

In an ideal world, I would expect everthing to just work. Instead, I get compiler errors about deleted operators on `json_value` due to it being a union of non trivial types.

### Compiler and operating system

N/A

### Library version

d4daaa897f48bf7bb2f96b46b84e49f32dd11daf

### Validation

- [X] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [X] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

**Comment by sjunges** [comment]
Just FYI: This https://github.com/moves-rwth/storm/commits/master/resources/3rdparty/modernjson/src diff shows some of the changes that were necessary to an earlier version of modernjson to support GMPs instead of doubles. 

We are in the progress of updating modernjson and applying the patch is a bit painful ;-)

**Comment by milestone-17** [comment]
I'm interested in working on this issue. I've read through the description and understand what needs to be done.

Could you please assign this issue to me? I'll submit a PR within [X days/weeks].

Thanks!

**Comment by nlohmann** [comment]
> I'm interested in working on this issue. I've read through the description and understand what needs to be done.

What should be done from your point of view?

**Comment by milestone-17** [comment]
Hi @eric-wieser, I've been looking into this issue and I'd like to work on a fix.

Root Cause:
The json_value union stores number types (number_integer_t, number_unsigned_t, number_float_t) by value. C++ unions cannot have members with non-trivial constructors/destructors, which prevents using types like mpz_class as number types.

Proposed Fix:
Store number types as pointers in the json_value union, exactly the same pattern already used for object_t*, array_t*, string_t*, and binary_t*. The existing create<T>() allocator infrastructure handles heap allocation and construction, and the destroy() method handles cleanup. This approach:

Is fully C++11 compatible
Reuses existing patterns in the codebase
Works with any number type (not just GMP)
Doesn't change the public API
Scope of changes:

json_value union: change number members to pointers
Use create<T>() for allocation in constructors
Add cleanup via allocator in destroy()
Update ~140 access points across the codebase (m_value.number_integer → dereference pattern)
The main change is systematic (all access patterns follow the same transformation), and the performance impact for the default case (trivial int64_t/uint64_t/double) is consistent with what's already accepted for string/object/array/binary types.

I've done a detailed analysis. If this approach sounds good, I'd be happy to implement it. I can start with a proof-of-concept using mpz_class as the test case.

This my first idea,maybe is not true,but please give me a chance. I promise to put in the time to get it fixed.

**Comment by milestone-17** [comment]
Hi @eric-wieser, I've been looking into this issue and I'd like to work on a fix
Could you give me your answer, whether you agree or not?
If you agree, I’ll be delighted and do my best to fix it.
If you disagree, please tell me where I fall short — I’ll still be happy to hear it

**Comment by milestone-17** [comment]
Hello,What you think about.
I am truely think that it is very subsitate to me.
Can you let me work it?
Thank you.

**Comment by nlohmann** [comment]
I am not convinced by the proposal. Moving the number type to pointers would slow down the behavior for the default number types. As long as we cannot make this change "free" for the current behavior, I think we cannot support GMP.

**Comment by milestone-17** [comment]
This is just my initial idea. If you assign this task to me, I will keep improving it and fix this issue thoroughly. I have internship experience at Baidu and possess strong professional capabilities.

**Comment by gregmarr** [comment]
I would think that it would be good to see if there is still interest in this before pursuing it.  If there is, then there should be a check for whether the number type is a primitive or object, and only convert the union to using a pointer if it is an object.  I am not sure what that would do to the rest of the code.  The comment at https://github.com/nlohmann/json/issues/3578#issuecomment-4467331170 mentions ~140 accesses, not sure what could be done with all of those.  I had some other thoughts about there being both a pointer and embedded storage, but I don't like that as it increases the storage size and adds an indirection to the common case, so it's not "free" for that.

**Comment by sjunges** [comment]
We would still be interested in this. I would think that given modern c++ options, we could resolve this at compile time rather than at run time?

**Comment by eric-wieser** [comment]
I don't think indirecting via a pointer is needed here; would redefining `json_value` via `std::variant` be an option?

**Comment by gregmarr** [comment]
>I don't think indirecting via a pointer is needed here; would redefining json_value via std::variant be an option?

No, that requires C++17, and this library still supports C++11.

Update: it could theoretically be done via a flag that would only be supported in C++17 and later and it would require another element in the ABI namespace macro.  This is not to say that such a change would be easy or clean, or that @nlohmann would accept such a change, that's just what the bare minimum would be.

**Comment by eric-wieser** [comment]
Fair, in that case my suggestion is "modify `json_value` to correctly handle object construction and destruction in the same way that `variant` does.

**Comment by nlohmann** [comment]
If we can cope with this at compile time and make sure the code is not becoming an even bigger ifdef-hell, then I'm good with this. Maybe there is a good way to abstract the 140 call sites.

**Comment by gregmarr** [comment]
@eric-wieser We already handle object construction and destruction for array, object, string, and binary, that's not the blocker.  What we don't want to do *for the base case* is to make numbers separately allocated objects.  So we want to find a way to have the number embedded in the union if it is a primitive type, exactly as it is now, and a separately allocated object if it is mpz or some other big number type, and then have as many of the other access points as possible not care which of the two it is, and at worst, the ones that do care call some function that handles it rather than having a local ifdef.

Maybe it's as simple as this:

```
#if numbers are primitive
        /// number (integer)
        number_integer_t number_integer;
        /// number (unsigned integer)
        number_unsigned_t number_unsigned;
        /// number (floating-point)
        number_float_t number_float;
        number_integer_t& integer() { return number_integer; }
        number_unsigned_t& unsigned() { return number_unsigned; }
        number_float_t& float() { return number_float; }
#else
        /// number (integer)
        number_integer_t *number_integer;
        /// number (unsigned integer)
        number_unsigned_t *number_unsigned;
        /// number (floating-point)
        number_float_t *number_float;
        number_integer_t& integer() { return *number_integer; }
        number_unsigned_t& unsigned() { return *number_unsigned; }
        number_float_t& float() { return *number_float; }
#endif
```

**Comment by eric-wieser** [comment]
> What we don't want to do _for the base case_ is to make numbers separately allocated objects

What I'm proposing is for the number type itself to never be an "allocated object", but instead be a C++ type with move and copy constructor and destructor, which is free to choose whether to allocate or not.

I believe gmp provides a gmpxx.h header that provides a value type hiding such allocation.

**Comment by sjunges** [comment]
What we do in our code above is summarized by the following main changes;

- json_value now stores the number_float as a pointer, if number_float_t is not trivial.
- Conversion operations between number representations are done using storm's utility functions.
- Type traits like std::is_scalar, std::is_trivial, and std::is_arithmetic_v are augmented by [custom definitions](https://github.com/stormchecker/storm/blob/master/resources/3rdparty/modernjson/include/nlohmann/storm_utility.hpp) within our namespace.

(see
https://github.com/stormchecker/storm/blob/master/resources/3rdparty/modernjson/README_STORM.md and hints on on how to get the precise diff [here](https://github.com/stormchecker/storm/blob/master/doc/update_resources.md).
) 

While I dont believe our approach is perfect, but if someone wants to look on what to change, this could serve as a starting point (and I while we are not C++11 compliant, I believe the essence of our implementation is).
